### DISCUSSION TITLE
"No job found" even though job returned

### DISCUSSION THREAD

**Discussion by vppetrov** [shinyorg/shiny#858] [Q&A] [answered]
I am registering and later starting a job with the following code:

```
var jobIdentifier = "MyJob";

jobManager.Register(new JobInfo(typeof(MyJobClass), jobIdentifier)
{
	RequiredInternetAccess = InternetAccess.Any,
	RunOnForeground = true,
	DeviceCharging = false,
	BatteryNotLow = true,
	Repeat = true
});
...
await jobManager.Run(jobIdentifier);
```

Sometimes this works perfectly but sometimes I get this message in the output console:

> Error: Error running job MyJob
>  System.ArgumentException: No job found named MyJob
>   at Shiny.Jobs.AbstractJobManager.Run (System.String jobName, System.Threading.CancellationToken cancelToken) [0x000ad] in <97c3aa17b66b4558a453b53c7c95957f>:0 

I changed my code to check that the job is registered before running it (even though I am sure the registration is started before the Run call):

```
if (await this.jobManager.GetJob(jobIdentifier) != null)
{
	await jobManager.Run(jobIdentifier);
}
```

But I get the same result, i.e. the GetJob() returns the JobInfo but afterward the error still appears.

Am I missing something?

**Comment by aritchie** [comment, marked as answer]
If you are calling Run immediately after the the Register, you'll need to await Register as it is async.

**Comment by vppetrov** [comment]
What is the recommended way to register the job on startup? The documentation container this FAQ:

> Q: When should I schedule my defined jobs? Should I do everytime I start the app.
> A: When your app starts (Xamarin Forms App.OnStart is a good place). It is a good practice to schedule every start. The framework will only update jobs if one already exists.

In the jobs sample in GitHub, the sample job is registered together with the services. In my case this is too early as other services are not registered yet.

So I want to register them in the app's OnStart method. However, if I make it async (in order to call "await Register"), the app fails to start on iOS with an exception because async methods are not allowed before the UI is shown.

**Reply by aritchie** [reply]
The docs are not complete. 

> In the jobs sample in GitHub, the sample job is registered together with the services. In my case this is too early as other services are not registered yet.

If you have services that your job depends on, you MUST register them with Shiny.  The background can startup without the rest of your app.  Registering jobs in OnStart of Xamarin Forms is not recommended.

**Reply by vppetrov** [reply]
> Registering jobs in OnStart of Xamarin Forms is not recommended.

That contradicts directly the FAQ from http://shinylib.net/jobs/faq. Am I reading the wrong documentation?

**Reply by aritchie** [reply]
The docs are not done and may never be done.  A lot of stuff is from my old open source project that I ported here.  Please remember that Shiny is not a product, it is a project I work on in my own time.  

Shiny was partially born from jobs and handles issues like dependencies better which is why I show job registration in the Shiny startup

**Reply by vppetrov** [reply]
Thanks Allan, appreciate your time and help!
